### DISCUSSION TITLE
How to render a component and have its source code syntax-highlighted in a `Pre` block?

### DISCUSSION THREAD

**Discussion by gpbl** [shuding/nextra#2167] [Q&A]
In my component page, I'd like to 1) display some source code as example, and 2) show how it renders – in the same page. Code samples arestored in a `../examples` directory, so that I can unit-test and reuse them.

In my page, I tried with the following code. However, it doesn't highlight the syntax of the `code` block with shiki.

```mdx

import firstExampleRaw from '!!raw-loader!../examples/first.tsx';
import FirstExample from '../examples/first.tsx';
import { Pre } from 'nextra/components';

## First Example

This code:

<Pre data-theme="tsx">{firstExampleRaw}</Pre>

renders:

<FirstExample />

```

Example: https://codesandbox.io/p/sandbox/smoosh-darkness-3dqm88?file=/pages/index.mdx

In Docusaurus, we have the [CodeBlock](https://docusaurus.io/docs/markdown-features/code-blocks#usage-in-jsx) component – which already highlight the source, while the Nextra's `Pre` does not.

Some related discussions I could find: https://github.com/shuding/nextra/discussions/1045, https://github.com/shuding/nextra/discussions/2119, https://github.com/shuding/nextra/issues/1195 (where [the linked doc](https://nextra.site/docs/guide/syntax-highlighting#with-dynamic-content) doesn't seem to help my case).

**Comment by iamhectorsosa** [comment]
This is exactly the issue we might face with this. @ericvalcik

**Comment by gpbl** [comment]
I could find an acceptable solution:

https://codesandbox.io/p/sandbox/73vs4k?file=/pages/index.mdx

<img width="868" alt="Screenshot 2023-08-22 at 11 22 43" src="https://github.com/shuding/nextra/assets/120693/9da242a1-87a1-4903-a072-527a6dddb598">

Basically, with the help of a custom component (`SourceCode`) and [getStaticProps](https://nextjs.org/docs/pages/building-your-application/data-fetching/get-static-props), I could use the shiki highlighter, the nextra built-in components and theme, to recreate the highlighted code blocks. Nice!

**Reply by iamhectorsosa** [reply]
Thanks for the solution! Worked like a charm! Didn't know that Nextra had these components exposed. If you don't mind my asking, is there a blog post explaining this? I'd love to do a write up on this with your blessing.

Saludos!

**Reply by gpbl** [reply]
> If you don't mind my asking, is there a blog post explaining this? I'd love to do a write up on this with your blessing.

Sure @iamhectorsosa, thanks - I'm sure it could help other people too. Would love to read your blog post, if you write one!

Is something worth a plugin or a package on npm? 

The code is in https://codesandbox.io/p/sandbox/73vs4k?file=/components/SourceCode.tsx and it is not unit-tested.

**Reply by iamhectorsosa** [reply]
Could be a nice idea, give me a couple of days to come up with the write up and maybe from there we'll have a better idea.

**Reply by preshonyee** [reply]
@gpbl @iamhectorsosa this solution works beautifully except that I can't get to highlight lines and substrings using this method. How would you suggest implementing the highlight line and substring?

**Reply by gpbl** [reply]
@preshonyee yeah, I recently discovered it. You can't even add line numbers. It seems these features are implemented in Nextra and/or the MDX parser. We should investigate better how these code blocks are rendered in Nextra or hope from some @shuding help. 

As alternative, I was thinking to "compile" the .mdx documents, before Nextra would parse them. When compiling, some placeholder is replaced by the source code like when we copy/paste it.

As last resort, we can do the copy/paste by ourself, thus giving up with this :)

**Reply by preshonyee** [reply]
@gpbl I got line numbers working by adding `data-line-numbers` prop to the Code component

```js return (
    <Pre
      data-theme="default"
      data-language="tsx"
      hasCopyCode
      filename={fileName}
    >
      <Code
        data-theme={theme}
        data-language="tsx"
        data-line-numbers
        dangerouslySetInnerHTML={{ __html: html }}
      />
    </Pre>
  );```

**Reply by preshonyee** [reply]
@gpbl also, I figured that each line of code is rendered in a `span` element within the `code` element. They have a default class of `line`. If you add `highlighted` as part of the class, the line gets highlighted. But I'm not sure how to map through each span node and update the class depending on the lineNumber passed in

<img width="688" alt="CleanShot 2023-10-24 at 15 33 58@2x" src="https://github.com/shuding/nextra/assets/19210093/f80cdeab-5329-41de-a879-311c4c96e8b1">

**Reply by preshonyee** [reply]
@gpbl I've got the highlighted lines working now.

**Reply by preshonyee** [reply]
@gpbl, @iamhectorsosa  here you go

## Here's how to get the lineNumbers and highlightLines working

Use this to parse the line numbers passed to the SourceCode component

```js
function parseHighlightLines(value: string): number[] {
  const parts = value.split(',');
  let lines: number[] = [];

  parts.forEach(part => {
    if (part.includes('-')) {
      const [start, end] = part.split('-').map(Number);
      lines = lines.concat(Array.from({ length: end - start + 1 }, (_, i) => start + i));
    } else {
      lines.push(Number(part));
    }
  });

  return lines;
}
```

Update the SourceCode props
```js
type SourceCodeProps = {
  /** The name of the file in the `docs/example` directory. */
  fileName: string;
  theme?: Theme;
  showLineNumbers?: boolean;
  /** The line numbers to highlight e.g highlightLines="1,3-4" */
  highlightLines?: string;
};
```

Apply the highlighted class to the necessary lines

```js
   const { fileName, theme, showLineNumbers, highlightLines } = props;
  
  // rest of your code

  // Get the lines to highlight
  const linesToHighlight = highlightLines ? parseHighlightLines(highlightLines) : [];

  // Apply the 'highlighted' class to the required lines
  let lineCount = 0;
  const highlightedHtml = html.replace(/<span class="line">/g, (match) => {
    lineCount++;
    return linesToHighlight.includes(lineCount) ? '<span class="line highlighted">' : match;
  });
  ```
  
  Add the `data-line-numbers` prop and update the `dangerouslySetInnerHTML` prop of the Code component
  
  ```jsx
  <Code
        data-theme={theme}
        data-language="tsx"
        {...(showLineNumbers ? { "data-line-numbers": true } : {})}
        dangerouslySetInnerHTML={{ __html: highlightedHtml }}
      />
```

Then in your `index.mdx` file

```mdx
### Example 2:

Displaying the source of `../examples/example2.tsx`.

<SourceCode fileName="example2.tsx" showLineNumbers={true} highlightLines="1,3-4" />
```

Full example is here https://codesandbox.io/p/sandbox/nextra-source-code-component-forked-zxj95t?file=%2Fpages%2Findex.mdx%3A41%2C13

**Reply by gpbl** [reply]
Thanks @preshonyee, terrific work! 

I like this final solution. Still it's lot of boilerplate code in the markdown. It would be nice if `SourceCode` could also render the example.

**Reply by preshonyee** [reply]
> Thanks @preshonyee, terrific work! 
> 
> 
> 
> I like this final solution. Still it's lot of boilerplate code in the markdown. It would be nice if `SourceCode` could also render the example.

@gpbl you mean like with using `react-live` or `react-runner` ??

**Reply by gpbl** [reply]
I was looking for a way to include examples without adding so much boilerplate to the MDX docs. I could go as far as:

````mdx
// Example of a markdown page

import { getSrc } from '@utils/getSrc';
import { SourceCode } from '@components/SourceCode';

export const getStaticProps = () => getSrc(["Button.tsx"]);

# Example 

Follow this example:

```include-example filename="Button.tsx"
Will show the source code and also render Button.tsx
```


We can use the SourceCode component to display the content of other files too:

<SourceCode src="example.css" showLineNumbers />

````

By tweaking the `theme.config.tsx` I could "add" a custom language `include-example` to the `pre` component, and then use the `SourceCode` component:

https://github.com/gpbl/react-day-picker/blob/next/docs/theme.config.tsx#L19-L34

`SourceCode` is also now able to render the example:

https://github.com/gpbl/react-day-picker/blob/next/docs/components/SourceCode/SourceCode.tsx#L130-L133.

There's a caveat: examples and example exports must be named in the same way. Then, they must be exported in a `examples/index.ts` file, so that we can import them via `import * from `examples`.

**Reply by preshonyee** [reply]
@gpbl, adding a custom language in the `theme.config.tsx` is a really elegant solution. I didn't realise this was possible. Is there a reason you didn't use `react-live-runner` or `react-live` for rendering the components instead??

I made something similar to your abstraction but using `react-live-runner`. This is what my markdown looks like now.

Although this fits my use-case well, but I've made a fork of your solution should in-case I need to switch directions.

```mdx
// Box.mdx

import {
  BoxKbdTable,
  BoxPlayground,
  BoxPropsTable,
} from '../components/examples/box';
import { getSourceCodeStaticProps } from '../utils/SourceCode';

export const getStaticProps = () =>
  getSourceCodeStaticProps(['box/default.tsx']);

# Box

<p className="component-description">A fundamental layout building block</p>

<BoxPlayground />

## API

### Layout Props

<BoxPropsTable />

## Accessibility

### Keyboard Interactions

<BoxKbdTable />
```

```tsx
// BoxPlayground.tsx

import React from 'react';

import Editor from '../../../utils/Editor';

import { BoxDemoDefault } from './default';

const examples = [
  {
    code: `render(<BoxDemoDefault />)`,
    scope: { BoxDemoDefault },
    fileName: 'box/default.tsx',
    title: 'default',
    description: 'this is the default Box component',
  },
];

export const BoxPlayground = () => {
  return <Editor examples={examples} showLineNumbers />;
};

```

```tsx 
// Editor.tsx

import React from 'react';

import Play from './play';
import { SourceCode } from './SourceCode';

type EditorProps = {
  examples: {
    code: string;
    scope: Record<string, unknown>;
    fileName: string;
    title: string;
    description: string;
  }[];
  showLineNumbers?: boolean;
  highlightLines?: string;
};

const Editor: React.FC<EditorProps> = ({
  examples,
  showLineNumbers,
  highlightLines,
}) => {
  return (
    <>
      {examples.map((example, index) => (
        <>
          <Play code={example.code} scope={example.scope} />
          <SourceCode
            fileName={example.fileName}
            showLineNumbers={showLineNumbers}
            highlightLines={highlightLines}
          />
        </>
      ))}
    </>
  );
};

export default Editor;
```

```tsx
// Play.tsx

import {
  LiveEditor,
  LiveError,
  LivePreview,
  LiveProvider,
} from 'react-live-runner';

type PlayProps = {
  code: string;
  scope: Record<string, unknown>;
};

const Play: React.FC<PlayProps> = ({ code, scope }) => {
  return (
      <>
        <LiveProvider code={code} scope={scope}>
          {/* <LiveEditor /> */}
          <LiveError />
          <LivePreview />
        </LiveProvider>
      </>
  );
};

export default Play;
```

**Reply by preshonyee** [reply]
@gpbl I also notice you're using Sandpack. I also use Sandpack too, but the bundling is still slow and doesn't make the playground ux really good.

I'm working on an abstraction that uses Sandpack under-the-hood to preview the components on a physical mobile device only via QR code. This way I can get real world mobile device testing versus mocking a viewport width in the playground.

**Reply by gpbl** [reply]
@preshonyee After using Sandpack, I am now in the opinion to leave out "live previews" or similar features from my docs. They are heavy and hard to configure, without many benefits. If somebody wants to play with the examples, we can always provide a CodeSandbox or try some of these new stuff in Github. 

Anyway, things are in progress, if you want to dig into the `next` branch of my project, here is the place: https://github.com/gpbl/react-day-picker/tree/next/docs 

Next step will be display the props from the jsdocs comments 🙀

**Reply by gpbl** [reply]
> adding a custom language in the `theme.config.tsx` is a really elegant solution. I didn't realise this was possible

Get familiar with the Nextra docs: https://nextra.site/docs/docs-theme/theme-configuration#mdx-components, they helped a lot!

**Reply by preshonyee** [reply]
@gpbl I also found a way to just render the code as a string with syntax highlighting and line numbers working without having to use the `fileName` or `filePath`

I've created a single file `SimpleSourceCode.tsx`. I kept getting errors related to loading wasm while trying to use multiple instances of shiki, so I'm using [shikiji](https://github.com/antfu/shikiji) instead of shiki since all grammars/themes/wasm are served as pure-ESM. 

```tsx
import React from 'react';
import { Code, Pre } from 'nextra/components';
import type { Lang, Theme } from 'shiki';
import { renderToHtml } from 'shiki';
import { getHighlighter } from 'shikiji';

type SourceCodeProps = {
  code?: string; // The string of code to render
  language?: Lang; // Language of the code
  theme?: Theme;
  showLineNumbers?: boolean;
  /** The line numbers to highlight e.g highlightLines="1,3-4" */
  highlightLines?: string;
};

export function SimpleSourceCode(props: SourceCodeProps) {
  const {
    code,
    language = 'tsx',
    theme,
    showLineNumbers,
    highlightLines,
  } = props;

  const [highlightedHtml, setHighlightedHtml] = React.useState('');

  React.useEffect(() => {
    async function highlight() {
      if (code) {
        const shiki = await getHighlighter({
          themes: ['min-light'],
          langs: [language],
        });

        const shikiHtml = shiki.codeToHtml(code, {
          lang: language,
          theme: 'min-light',
        });
        setHighlightedHtml(shikiHtml);
      }
    }
    highlight();
  }, [code, language]);

  const html = highlightedHtml
    // Remove out HTML components we replace with nextra components
    .replaceAll('</code>', '')
    .replaceAll('<code>', '')
    .replace(/<pre [^>]*>/gi, '')
    .replace('</pre>', '');
  // For some reasons, when doing this empty lines are lost
  // .replaceAll(
  //   '<span class="line"></span>',
  //   '<span class="line">\n</span>',
  //   'gi',
  // );

  // Highlight lines logic
  const linesToHighlight = highlightLines
    ? parseHighlightLines(highlightLines)
    : [];
  let lineCount = 0;
  const highlightedWithLines = html.replace(/<span class="line">/g, (match) => {
    lineCount++;
    return linesToHighlight.includes(lineCount)
      ? '<span class="line highlighted">'
      : match;
  });

  return (
    <Pre data-theme="default" data-language={language} hasCopyCode>
      <Code
        data-theme={theme}
        data-language={language}
        {...(showLineNumbers ? { 'data-line-numbers': true } : {})}
        dangerouslySetInnerHTML={{ __html: highlightedWithLines }}
      />
    </Pre>
  );
}

function parseHighlightLines(value: string): number[] {
  const parts = value.split(',');
  let lines: number[] = [];

  parts.forEach((part) => {
    if (part.includes('-')) {
      const [start, end] = part.split('-').map(Number);
      lines = lines.concat(
        Array.from({ length: end - start + 1 }, (_, i) => start + i),
      );
    } else {
      lines.push(Number(part));
    }
  });

  return lines;
}
```

You can use `shiki.codeToThemedTokens` instead of `shiki.codeToHtml`. Both do the same work at the end

```tsx
  const [highlightedHtml, setHighlightedHtml] = React.useState('');

  React.useEffect(() => {
    async function highlight() {
      if (code) {
        const shiki = await getHighlighter({
          themes: ['min-light'],
          langs: [language],
        });

        const shikiCode = shiki.codeToThemedTokens(code, {
          lang: language,
          theme: 'min-light',
        });
        const shikiHtml = renderToHtml(shikiCode);
        setHighlightedHtml(shikiHtml);
      }
    }
    highlight();
  }, [code, language]);
```

This is what my markdown looks like, just that I can't get prettier formatting working nicely for template string literals. I can also use the `SimpleSourceCode` in a `tsx` file if I want.

```mdx
import { Callout } from 'nextra/components';

import { SimpleSourceCode } from '../utils/SimpleSourceCode';

# Codebase setup

- VSCode settings for productivity
- VSCode tips and trick
- VSCode extensions

<Callout type="default" emoji="️🚫">
  This is a dangerous feature that can cause everything to explode.
</Callout>

<SimpleSourceCode
        code={`import React from 'react';
import { Box } from '../../../ds/design-system/build-sandpack';
import PlaceholderBox from '../../DecoratedBox';

export const BoxDemoDefault = () => {
  return (
    <Box
    css={\`
      background-color: #f3f4f6;
      border-radius: 8px;
      padding: 1rem;
    \`}>
      <PlaceholderBox />
    </Box>
  );
};`}
        language="tsx"
        showLineNumbers={true}
        highlightLines="3"
      />

```

**Reply by gpbl** [reply]
Good find with https://github.com/antfu/shikiji thanks for sharing. 

Could you, with this last solution, import the raw content of the source? E.g. something like

````mdx
import boxExample from '!!raw-loader!../boxExample.tsx';

## Markdown example

Source code:

<SimpleSourceCode code={boxExample}> ;
````

I could try it by myself :) I need to focus on other things. Thanks for sharing your solution, we may end up with a plugin for nextra?

**Reply by preshonyee** [reply]
@gpbl got it working by doing this

```tsx
export const Example = () => {
  const [code, setCode] = React.useState('');

  const getSomething = async () => {
    const raw = await import(`!!raw-loader!./Kbd.tsx`);
    return raw;
  };

  React.useEffect(() => {
    getSomething().then((res) => {
      setCode(res.default);
    });
  }, []);

  return (
    <Box>
      <SimpleSourceCode
        code={code}
        language="tsx"
        showLineNumbers={true}
        highlightLines="3"
      />
    </Box>
  );
};
```

**Reply by preshonyee** [reply]
@gpbl  Do you know of a similar way to load the raw content of the source of a component without loading the whole content of the file?

Say for example I have multiple exports in one file, is there a way to load the raw source of each components in the same file, without having to make separate files?

```tsx
import { Comp1, Comp2 } from `!!raw-loader!./example.tsx`);
```

**Reply by gpbl** [reply]
@preshonyee AFAIK you cannot. You can however import the whole content of the file and parse it somehow to "extract" the functions.

**Comment by crguezl** [comment]
The comments above refer to Nextra 2. This is a comment for Nextra 3 users. 

Using `RemoteContent`, `buildDynamicMDX` and `getStaticProps` or `getServerSideProps` you can  manage to not only render code  from a file but any arbitrary mdx!

See an example at my tutorial [Dynamically read and process a MDX File](https://ull-mii-sytws.github.io/temas/web/nextra/dynamic-readmdxfile.html#dynamically-read-and-process-a-mdx-file)
